### PR TITLE
Hostnames

### PR DESCRIPTION
**Pull request by sthulb**
This work sets the machine name in the Cloud API or Hypervisor. As well
as setting the hostname inside the VM.

I've added the machine name to the NewDriver func to allow for
identification by Cloud APIs and for the driver package.

Each driver will attempt to set the hostname for the VM, except for
Azure, which sets it based on the DNS name.

Refs: #59

### CODE REVIEW

File: drivers/virtualbox/virtualbox.go [hunk 1]
```diff
@@ -363,6 +363,14 @@ func (d *Driver) Create() error {
 		return err
 	}
 
+	cmd, err = d.GetSSHCommand(fmt.Sprintf("sudo hostname %s", d.MachineName))
```
Comment (ehazlett) [inline drivers/virtualbox/virtualbox.go]: I don't believe this will persist reboots will it (need to update `/etc/hostname` at least).

Comment (bfirsh) [inline drivers/virtualbox/virtualbox.go]: Does that persist on boot2docker? Perhaps we don't bother setting the hostname on boot2docker... I think the big win from this pull request is the correct name showing up in your cloud provider's admin (e.g. list of instances on EC2), not the hostname when you log in.

Comment (sthulb) [inline drivers/virtualbox/virtualbox.go]: You're right, it won't persist reboots. However one needs to update `/usr/local/etc/hostname` for it to persist on a b2d image.

Comment (sthulb) [inline drivers/virtualbox/virtualbox.go]: @bfirsh I'm going for naming all of the things.

**Comment by bfirsh** [comment]
Thanks for this @sthulb. The bit I'm really interested in is setting the correct name on cloud providers. E.g. on Digital Ocean, removing the concept of `DropletName` and just making the name of the droplet the name of the host.

**Comment by sthulb** [comment]
EC2 support is kind of there, I can tag an instance with the machine ID, just the internal hostname remaining.

**Comment by sthulb** [comment]
I moved how we set the hostname to the `driver.Create` func since different drivers need the hostname func to be called at different times.

File: drivers/amazonec2/amazonec2.go [hunk 2]
```diff
@@ -225,9 +228,32 @@ func (d *Driver) Create() error {
 		return err
 	}
 
+	log.Infof("Settings tags for instance")
```
Comment (ehazlett) [inline drivers/amazonec2/amazonec2.go]: This should probably be Debug

**Comment by ehazlett** [comment]
@sthulb i like this so far -- good work!

File: drivers/amazonec2/amazonec2.go (hunk 2 above)
Comment (sthulb) [inline drivers/amazonec2/amazonec2.go]: If you prefer :)

**Comment by bfirsh** [comment]
Great stuff. Since a driver is always attached to a machine, perhaps the name should be passed to `NewDriver`?

**Comment by sthulb** [comment]
@bfirsh It's something that should be considered when we need it in other places, but for now, I think leaving it on `Create(name string)` seems ok.

**Comment by sthulb** [comment]
@bfirsh I take back my comment... azure kind of forces your idea.

**Comment by sthulb** [comment]
@bfirsh @ehazlett I think it's finished. someone should perhaps test the Azure & Google drivers

**Comment by ehazlett** [comment]
awesome!  thanks -- I will check it out

**Comment by ehazlett** [comment]
I'm getting "sudo: unable to resolve host test-ec2" with `amazonec2`.  The hostname isn't added to `/etc/hosts`.  You may need to silence this for the first sudo to add it but I think it would be good otherwise it will spit this out for every `sudo` command.

**Comment by sthulb** [comment]
I'll alter the PR to add to the hosts file also.

**Comment by ehazlett** [comment]
@sthulb thanks

**Comment by ehazlett** [comment]
@sthulb i'm also getting the following with azure:

```
FATA[0004] A hosted service with the specified name already exists.
```

It's not happening in master.

**Comment by sthulb** [comment]
I've fixed the sudo thing.

**Comment by sthulb** [comment]
So Azure likes globally unique names... so the best we can do is: `foo-<timestamp>` for azure, which sucks. Unless someone knows differently?

**Comment by jeffmendoza** [comment]
In Azure, your VM sits inside a "Cloud Service." The cloud service must have a globally unique name, as it's used for DNS: <name>.cloudapp.net. The VM can have any name, and that will be the hostname of the VM.

**Comment by ehazlett** [comment]
@sthulb i think naming with a unique/timestamp is fine.  it's just the cloud service and not the actual VM hostname which is what we want to name anyway.

**Comment by ehazlett** [comment]
@sthulb sorry to be nitpicky but I'm still getting a couple `sudo: unable to resolve host test-ec2`.  We may need to `> /dev/null 2>&1` or something to hide the initial set.  wdyt?

**Comment by sthulb** [comment]
That's fine, ... I just need to track down the origin of it...

**Comment by ehazlett** [comment]
@sthulb Hmm I keep getting this error with EC2:

```
Error creating host: Error decoding error response: Error decoding error response: http: read on closed response body
```

Have you ran into that?

**Comment by sthulb** [comment]
Not at all.

**Comment by ehazlett** [comment]
Hmm ok -- may be AWS.  I'll try again -- thx

**Comment by sthulb** [comment]
Perhaps we can get @bfirsh try it too.

**Comment by ehazlett** [comment]
@sthulb can you do a fresh build and confirm it's not doing it?  It's not doing it to me in master.

**Comment by ehazlett** [comment]
@sthulb +1 /cc @bfirsh

File: drivers/digitalocean/digitalocean.go [hunk 3]
```diff
@@ -147,9 +145,24 @@ func (d *Driver) Create() error {
 		return err
 	}
 
+	log.Debugf("Setting hostname: %s", d.MachineName)
+	cmd, err := d.GetSSHCommand(fmt.Sprintf(
+		"echo \"127.0.0.1 %s\" | sudo tee -a /etc/hosts && sudo hostname %s && echo \"%s\" | sudo tee /etc/hostname",
```
Comment (bfirsh) [inline drivers/digitalocean/digitalocean.go]: Doesn't Digital Ocean set the hostname to be the droplet name?

**Comment by sthulb** [comment]
@ehazlett I think I've figured out why you get it... By any chance, do you already have a security group with the same group name?

**Comment by ehazlett** [comment]
yep.  looks like it was left over from a previous run.  i will remove and
test.  thanks!

**Comment by ehazlett** [comment]
LGTM

**Pull request merged**